### ISSUE TITLE
Providing insight into the structure of data definitions

### ISSUE THREAD

**Issue by benbc** [opensafely-core/databuilder#677]
This is a proposal for shore leave work.

# Context

The query model in Data Builder gives us a formal, machine-inspectable form of data definitions that researchers write. At the moment all we use that for is translating it into SQL, but we could do all sorts of other things with it as well.

# Proposal

## Why?

* To give us insight into the structure of data definitions that researchers write.
* To give researchers insight into the results of the code that they're writing.

I don't have any real goals for what we might do about these insights -- it's all rather exploratory.

## What? 

* "Raw" query model is very hard to read. Is there another representation we can come up with? Maybe useful for debugging.
* See the structure of queries.
* Spot duplication within the query graphs.
* Are there subgraphs that can be pruned or simplified without changing the results?

## How?

* Explore clearer text representations than the code literals that we currently use.
* Explore graphical display of the query graph. Try Graphviz. Any others?
* Look for representations that are guaranteed to merge identical subgraphs.
* Produce statistics representing the structure of the graph.

**Comment by evansd** [comment]
If, as part of this work, you ended up looking at:
 * https://github.com/opensafely-core/databuilder/issues/573

I would not be altogether displeased :)

**Comment by benbc** [comment]
@evansd Yes, I looked at that as something to work on in itself. It's sort of related, isn't it? I think one possible (but not super-likely) outcome of doing this analysis may be to identify transformations that we want to do.

**Comment by evansd** [comment]
Yeah, it's probably too big a thing to tackle in passing. I just wanted to plant the seed in case you had a burst of inspiration while working on this and realised it was all incredibly simple!
